pyiceberg/io/pyarrow.py

```diff
@@ -698,77 +708,143 @@ def before_field(self, field: pa.Field) -> None:
-            field_type = field_results[i]
-            if field_type is not None and field_id is not None:
-                fields.append(NestedField(field_id, field.name, field_type, required=not field.nullable, doc=field_doc))
-        return fields
-
-    def schema(self, schema: pa.Schema, field_results: List[Optional[IcebergType]]) -> Schema:
-        return Schema(*self._convert_fields(schema, field_results))
-
-    def struct(self, struct: pa.StructType, field_results: List[Optional[IcebergType]]) -> IcebergType:
-        return StructType(*self._convert_fields(struct, field_results))
-
-    def list(self, list_type: pa.ListType, element_result: Optional[IcebergType]) -> Optional[IcebergType]:
+class _HasIds(PyArrowSchemaVisitor[bool]):
+    def schema(self, schema: pa.Schema, struct_result: bool) -> bool:
+        return struct_result
+
+    def struct(self, struct: pa.StructType, field_results: List[bool]) -> bool:
+        return all(field_results)
+
+    def field(self, field: pa.Field, field_result: bool) -> bool:
+        return all([_get_field_id(field) is not None, field_result])
+
+    def list(self, list_type: pa.ListType, element_result: bool) -> bool:
         element_field = list_type.value_field
         element_id = _get_field_id(element_field)
-        if element_result is not None and element_id is not None:
-            return ListType(element_id, element_result, element_required=not element_field.nullable)
-        return None
+        return element_result and element_id is not None
 
-    def map(
-        self, map_type: pa.MapType, key_result: Optional[IcebergType], value_result: Optional[IcebergType]
-    ) -> Optional[IcebergType]:
+    def map(self, map_type: pa.MapType, key_result: bool, value_result: bool) -> bool:
         key_field = map_type.key_field
         key_id = _get_field_id(key_field)
         value_field = map_type.item_field
         value_id = _get_field_id(value_field)
-        if key_result is not None and value_result is not None and key_id is not None and value_id is not None:
-            return MapType(key_id, key_result, value_id, value_result, value_required=not value_field.nullable)
-        return None
+        return all([key_id is not None, value_id is not None, key_result, value_result])
+
+    def primitive(self, primitive: pa.DataType) -> bool:
+        return True
 
-    def primitive(self, primitive: pa.DataType) -> IcebergType:
+
+class _ConvertToIceberg(PyArrowSchemaVisitor[Union[IcebergType, Schema]]):
+    """Converts PyArrowSchema to Iceberg Schema. Applies the IDs from name_mapping if provided."""
+
+    field_names: List[str]
+    name_mapping: Optional[NameMapping]
+
+    def __init__(self, name_mapping: Optional[NameMapping] = None) -> None:
+        self.field_names = []
+        self.name_mapping = name_mapping
+
+    def _current_path(self) -> str:
+        return ".".join(self.field_names)
+
+    def _get_field_id(self, field: pa.Field) -> int:
```

Comment:
@HonahX @Fokko - I wanted to get your thoughts on an idea.

In spark, we can currently [create an Iceberg table](https://iceberg.apache.org/docs/latest/spark-writes/#creating-tables) without having to explicitly specify the schema and instead relying on spark-iceberg's conversion of spark dataframe schema to iceberg schema.

It looks like we currently have no way of converting a Pyarrow Table schema to an Iceberg Schema that will allow a user to construct a PyIceberg easily. Right now, a user has to manually construct a PyIceberg schema from scratch.

Would it make sense to pass an optional argument "**assign_ids**" to _ConvertToIceberg class where the user can ask the class to assign field IDs in order, instead of using the attached metadata field_id or the name_mapping?

This optional argument will be used by pyarrow_to_schema, only when a user wants to parse a pyarrow schema to PyIceberg schema for the purpose of creating a new table, and should not be used when reading an Iceberg Table.

Reply:
Thanks for raising this.

For the Write support what we do is: Convert the Arrow schema to Iceberg. Create a new table from it where it will [get fresh IDs assigned](https://github.com/apache/iceberg-python/blob/4593208f67e1101a94059a93bfb13f609ac2ad80/pyiceberg/schema.py#L1216-L1218), and then you can use that schema to write the dataframe to the table.

> Would it make sense to pass an optional argument "assign_ids" to _ConvertToIceberg class where the user can ask the class to assign field IDs in order, instead of using the attached metadata field_id or the name_mapping?

I don't think that's a good idea, and this is mostly from the learnings that we had from Java. For example:

```
1: firstname string
2: lastname string
```

And you add a field called `title`:

```
1: title string
2: firstname string
3: lastname string
```

This way we re-assign the fields, and we could only add fields at the end. But this gets complex with nested fields, because they need to know about pre-order traversal.

Can you elaborate more on the use-case? A workaround is to create a name-mapping, and then pass the name mapping to the visitor. After this PR, we probably want to add the [UnionByNameVisitor](https://github.com/apache/iceberg/blob/main/core/src/main/java/org/apache/iceberg/schema/UnionByNameVisitor.java) as well. This is important when we want to do the schema evolution.

Related, in https://github.com/apache/iceberg-python/pull/252 I'm working on adding the field-IDs to the Arrow schema. This should it also make it easier to perform the union, in case you want to write to an existing table, but with a new column. This PR is blocked on the new Arrow release.

Reply:
Hi @Fokko thank you for your response. I think the use case I'm concerned with, is similar to the one that @robtandy has brought up in the [open Write Support PR](https://github.com/apache/iceberg-python/pull/41/files#r1440972324). If a user wants to create a new Iceberg table when they have an in memory Arrow table available. What we are proposing here sounds like:
1. Get the Arrow Schema from the table (this does not have field.metadata or field_ids)
2. Convert the Arrow Schema to PyIceberg Schema (using pyarrow_to_schema, which uses _ConvertToIceberg)
3. Assign fresh IDs to the PyIceberg Schema
4. Use the PyIceberg Schema in create_table call

I think right now, (2) would fail, because _ConvertToIceberg requires that the field_ids or field_metadata be present in the arrow fields, or that we provide a name_mapping. From my understanding, _ConvertToIceberg is the only schema conversion class that takes an Arrow schema as the input and outputs a PyIceberg schema. I think asking the user to provide the name_mapping also leads to the same problem, because the name mapping generating function also requires PyIceberg Schema as the input.

Reply:
I think @syun64 is right that currently we do not have a convenient way to convert an in memory arrow table to a new iceberg table.  I'm +1 on adding a utility method to extract the iceberg schema from a pyarrow table, similar to the existing [SparkSchemaUtil.schemaForTable](https://github.com/apache/iceberg/blob/31e31fd819c846f49d2bd459b8bfadfdc3c2bc3a/spark/v3.5/spark/src/main/java/org/apache/iceberg/spark/SparkSchemaUtil.java#L51-L65). A similar use case is in table migration: we convert Spark or Delta Lake schemas to new iceberg schemas with fresh ids for use in `createTableTransaction`

I think it's crucial for the method's docstring to specify that this utility should be employed exclusively when creating a new iceberg table. It might be better to make this a distinct method (may be named as `new_schema_for_table`) separate from `pyarrow_to_schema` to avoid confusion. As for the implementation details, I think we can introduce a new visitor (`_ConvertToIcebergWithFreshIds`) for this. Do you think this sounds reasonable?

Reply:
> It might be better to make this a distinct method (may be named as new_schema_for_table) separate from pyarrow_to_schema to avoid confusion

I'm 100% in agreement with this.

> As for the implementation details, I think we can introduce a new visitor (_ConvertToIcebergWithFreshIds) for this.

That makes sense as well - I think we can use a simpler version of _ConvertToIceberg without all the enhancements related to name_mapping support for this task.

Thank you for your input @HonahX !